File: src/Python/sync_vote_activities.py
"""Sync vote activities from Koios + IPFS → PostgreSQL ga_* tables.

Koios: GET /api/v1/proposal_votes?_proposal_id={id}
IPFS: Fetch metadata for comment
Target: ga_* tables (one per proposal) in PostgreSQL
Reads proposals with activities_table_name from proposals table.
"""

import sys
import time

from config import GA_TABLE_COLUMNS, BATCH_SIZE, API_DELAY
from helpers import (
    get_logger, check_env, koios_get, fetch_ipfs_metadata,
    pg_connect, pg_upsert_batch, pg_row_count,
    now_iso, gen_uuid, pg_query,
)


def sync_vote_activities(logger=None, only_active=False):
    if logger is None:
        logger = get_logger()

    logger.info("=== Sync: vote_activities (Koios+IPFS → ga_* tables) ===")

    # 1. Get proposals with activities_table_name from PostgreSQL
    conn = pg_connect()
    try:
        if only_active:
            proposals_res = pg_query(
                conn,
                "SELECT proposal_id, activities_table_name FROM proposals "
                "WHERE activities_table_name IS NOT NULL "
                "AND status IN ('voting','active') ORDER BY proposal_id"
            )
        else:
            proposals_res = pg_query(
                conn,
                "SELECT proposal_id, activities_table_name FROM proposals "
                "WHERE activities_table_name IS NOT NULL ORDER BY proposal_id"
            )
        proposals = [(r[0], r[1]) for r in proposals_res]
    finally:
        conn.close()
    # Skip test proposals
    proposals = [(pid, tbl) for pid, tbl in proposals if not pid.startswith("final_verify_v3")]

    logger.info(f"[vote_activities] Found {len(proposals)} proposals with ga_* tables")

    # 2. Process each proposal
    total_votes = 0
    errors = 0
    total_skipped = 0

    for idx, (pid, table_name) in enumerate(proposals):
        logger.info(f"\n  [{idx+1}/{len(proposals)}] {pid} → {table_name}")

        conn = pg_connect()
        try:
            # Fetch votes from Koios
            votes = koios_get("proposal_votes", params={"_proposal_id": pid})
            if not votes:
                logger.info(f"    No votes found")
                continue

            logger.info(f"    Got {len(votes)} votes from Koios")

            # Check if there are new votes (compare row count)
            count_res = pg_query(conn, f'SELECT count(*) FROM "{table_name}"')
            current_count = count_res[0][0] if count_res else 0

            # Build a set of (voter_id, block_time) that already have comments
            existing_with_comments = set()
            if current_count > 0:
                existing_res = pg_query(
                    conn,
                    f'SELECT voter_id, block_time FROM "{table_name}" '
                    f'WHERE comment IS NOT NULL AND comment != \'\''
                )
                for r in existing_res:
                    existing_with_comments.add((r[0], str(r[1])))

            if current_count == len(votes) and len(existing_with_comments) >= len(votes):
                logger.info(f"    Already up-to-date ({current_count} rows, all have comments), skipping")
                continue

            new_count = len(votes) - current_count
            if current_count < len(votes):
                logger.info(f"    {new_count} new votes to upsert (DB={current_count}, Koios={len(votes)})")
            else:
                logger.info(f"    Upserting {len(votes)} votes (DB={current_count}, Koios={len(votes)})")
            if existing_with_comments:
                logger.info(f"    {len(existing_with_comments)} votes already have comments (will skip IPFS)")

            # Process votes
            rows = []
            skipped = 0
            for v in votes:
                block_time = v.get("block_time")
                if isinstance(block_time, (int, float)):
                    bt_iso = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime(block_time))
                else:
                    bt_iso = str(block_time)

                voter_id = v.get("voter_id")
                meta_url = v.get("meta_url")

                # Skip IPFS fetch if vote already has comment in DB
                if (voter_id, bt_iso) in existing_with_comments:
                    skipped += 1
                    continue

                # Fetch IPFS metadata for comment
                comment = ""
                if meta_url:
                    comment = fetch_ipfs_metadata(meta_url, voter_role=v.get("voter_role"))

                row = {
                    "id": gen_uuid(),
                    "block_time": bt_iso,
                    "voter_role": v.get("voter_role"),
                    "voter_id": voter_id,
                    "vote": v.get("vote"),
                    "meta_url": meta_url,
                    "comment": comment,
                    "processed_at": now_iso(),
                    "created_at": now_iso(),
                    "updated_at": now_iso(),
                }
                rows.append(row)

            # Insert in batches
            inserted = 0
            for i in range(0, len(rows), BATCH_SIZE):
                batch = rows[i:i + BATCH_SIZE]
                try:
                    batch_cols = [c for c in GA_TABLE_COLUMNS if all(c in row for row in batch)]
                    count = pg_upsert_batch(
                        conn, table_name, batch_cols, batch,
                        conflict_cols=["voter_id", "block_time"]
                    )
                    inserted += count
                except Exception as e:
                    logger.error(f"    Insert error at batch {i}: {e}")

            total_votes += inserted
            total_skipped += skipped
            logger.info(f"    Inserted {inserted}/{len(rows)}, skipped {skipped}")

        except Exception as e:
            errors += 1
            logger.error(f"    Error: {e}")
        finally:
            conn.close()

        time.sleep(API_DELAY)

    logger.info(f"\n[vote_activities] Total votes inserted: {total_votes}, skipped: {total_skipped}, errors: {errors}")
    return total_votes


if __name__ == "__main__":
    check_env()
    only_active = "--active-only" in sys.argv
    sync_vote_activities(only_active=only_active)

File: src/JavaScript/sync_vote_activities.js
// sync_vote_activities.js
// Sync vote activities from Koios + IPFS → PostgreSQL ga_* tables
// Equivalent to sync_vote_activities.py
//
// Koios: GET /api/v1/proposal_votes?_proposal_id={id}
// IPFS: Fetch metadata for comment
// Target: ga_* tables (one per proposal) in PostgreSQL
// Reads proposals with activities_table_name from proposals table.

require('dotenv').config();
const fetch = require('node-fetch');
const crypto = require('crypto');
const { pgConnect, pgUpsertBatch, fetchIpfsMetadata, info, warn, error } = require('./helpers');
const { ga_table: GA_TABLE_COLUMNS } = require('./config');

const DATABASE_URL = process.env.DATABASE_URL;
if (!DATABASE_URL) {
  console.error('❌ Missing DATABASE_URL in .env');
  process.exit(1);
}

const KOIOS_BASE = 'https://api.koios.rest/api/v1';
const BATCH_SIZE = 25;
const API_DELAY = 500; // ms

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

function genUuid() {
  return crypto.randomUUID();
}

function nowIso() {
  return new Date().toISOString();
}

function toIso(blockTime) {
  if (typeof blockTime === 'number') {
    return new Date(blockTime * 1000).toISOString();
  }
  return String(blockTime);
}

async function koiosGet(endpoint, params = {}) {
  const url = new URL(`${KOIOS_BASE}/${endpoint}`);
  for (const [k, v] of Object.entries(params)) {
    url.searchParams.set(k, v);
  }
  const res = await fetch(url.toString(), {
    headers: { Accept: 'application/json' },
  });
  if (!res.ok) {
    throw new Error(`Koios ${endpoint} → HTTP ${res.status}`);
  }
  return res.json();
}

async function syncVoteActivities({ onlyActive = false } = {}) {
  info('=== Sync: vote_activities (Koios+IPFS → ga_* tables) ===');

  const client = await pgConnect();

  try {
    // 1. Get proposals with activities_table_name from PostgreSQL
    const q = onlyActive
      ? `SELECT proposal_id, activities_table_name FROM proposals
         WHERE activities_table_name IS NOT NULL
         AND status IN ('voting','active') ORDER BY proposal_id`
      : `SELECT proposal_id, activities_table_name FROM proposals
         WHERE activities_table_name IS NOT NULL ORDER BY proposal_id`;
    const res = await client.query(q);
    const proposals = res.rows
      .map(r => [r.proposal_id, r.activities_table_name])
      .filter(([pid]) => !String(pid).startsWith('final_verify_v3'));

    info(`[vote_activities] Found ${proposals.length} proposals with ga_* tables`);

    let totalVotes = 0;
    let errors = 0;
    let totalSkipped = 0;

    for (let idx = 0; idx < proposals.length; idx++) {
      const [pid, tableName] = proposals[idx];
      info(`\n  [${idx + 1}/${proposals.length}] ${pid} → ${tableName}`);

      try {
        // 2. Fetch votes from Koios
        const votes = await koiosGet('proposal_votes', { _proposal_id: pid });
        if (!votes || votes.length === 0) {
          info('    No votes found');
          continue;
        }
        info(`    Got ${votes.length} votes from Koios`);

        // 3. Check current row count + votes that already have comments
        const countRes = await client.query(`SELECT count(*)::int AS cnt FROM "${tableName}"`);
        const currentCount = countRes.rows[0].cnt;

        const existingWithComments = new Set();
        if (currentCount > 0) {
          const cRes = await client.query(
            `SELECT voter_id, block_time FROM "${tableName}" WHERE comment IS NOT NULL AND comment != ''`
          );
          for (const r of cRes.rows) {
            existingWithComments.add(`${r.voter_id}|${String(r.block_time)}`);
          }
        }

        if (currentCount === votes.length && existingWithComments.size >= votes.length) {
          info(`    Already up-to-date (${currentCount} rows, all have comments), skipping`);
          continue;
        }

        const newCount = votes.length - currentCount;
        if (currentCount < votes.length) {
          info(`    ${newCount} new votes to upsert (DB=${currentCount}, Koios=${votes.length})`);
        } else {
          info(`    Upserting ${votes.length} votes (DB=${currentCount}, Koios=${votes.length})`);
        }
        if (existingWithComments.size) {
          info(`    ${existingWithComments.size} votes already have comments (will skip IPFS)`);
        }

        // 4. Process votes
        const rows = [];
        let skipped = 0;
        for (const v of votes) {
          const btIso = toIso(v.block_time);
          const voterId = v.voter_id;
          const metaUrl = v.meta_url;

          // Skip IPFS fetch if vote already has comment in DB
          if (existingWithComments.has(`${voterId}|${btIso}`)) {
            skipped++;
            continue;
          }

          let comment = '';
          if (metaUrl) {
            comment = (await fetchIpfsMetadata(metaUrl, v.voter_role)) || '';
          }

          rows.push({
            id: genUuid(),
            block_time: btIso,
            voter_role: v.voter_role,
            voter_id: voterId,
            vote: v.vote,
            meta_url: metaUrl,
            comment,
            processed_at: nowIso(),
            created_at: nowIso(),
            updated_at: nowIso(),
          });
        }

        // 5. Insert in batches
        let inserted = 0;
        for (let i = 0; i < rows.length; i += BATCH_SIZE) {
          const batch = rows.slice(i, i + BATCH_SIZE);
          try {
            const batchCols = GA_TABLE_COLUMNS.filter(c => batch.every(r => c in r));
            inserted += await pgUpsertBatch(client, tableName, batchCols, batch, {
              conflict_cols: ['voter_id', 'block_time'],
              preserve_cols: ['id'],
              do_update: true,
            });
          } catch (e) {
            error(`    Insert error at batch ${i}: ${e.message}`);
          }
        }

        totalVotes += inserted;
        totalSkipped += skipped;
        info(`    Inserted ${inserted}/${rows.length}, skipped ${skipped}`);
      } catch (e) {
        errors++;
        error(`    Error: ${e.message}`);
      }

      await sleep(API_DELAY);
    }

    info(
      `\n[vote_activities] Total votes inserted: ${totalVotes}, skipped: ${totalSkipped}, errors: ${errors}`
    );
    return totalVotes;
  } finally {
    await client.end();
  }
}

if (require.main === module) {
  const onlyActive = process.argv.includes('--active-only');
  syncVoteActivities({ onlyActive }).catch(err => {
    console.error(err);
    process.exit(1);
  });
}

module.exports = { syncVoteActivities };